File: array_search/search_binary/src/BinarySearchDemo.cs
// C#
// 配列の検索: 二分探索 (Binary Search)

using System;
using System.Collections.Generic;

namespace BinarySearchDemo
{
    class ArrayData
    {
        private List<int> _data = new List<int>();

        public List<int> Get()
        {
            return _data;
        }

        public bool Set(List<int> data)
        {
            _data = data;
            return true;
        }

        public int Search(int target)
        {
            int left = 0;
            int right = _data.Count - 1;

            while (left <= right)
            {
                int mid = (left + right) / 2;

                // 中央の要素が目標値と一致
                if (_data[mid] == target)
                {
                    return mid;
                }
                // 中央の要素が目標値より小さい場合、右半分を探索
                else if (_data[mid] < target)
                {
                    left = mid + 1;
                }
                // 中央の要素が目標値より大きい場合、左半分を探索
                else
                {
                    right = mid - 1;
                }
            }

            // 目標値が見つからない場合
            return -1;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("BinarySearch TEST -----> start");

            Console.WriteLine("\nnew");
            ArrayData arrayData = new ArrayData();
            List<int> input = new List<int> { 1, 3, 5, 7, 9, 11, 13, 15, 17, 19 };
            arrayData.Set(input);
            Console.WriteLine($"  現在のデータ: [{string.Join(", ", arrayData.Get())}]");

            Console.WriteLine("\nsearch");
            int searchValue = 7;
            Console.WriteLine($"  入力値: {searchValue}");
            int output = arrayData.Search(searchValue);
            Console.WriteLine($"  出力値: {output}");

            Console.WriteLine("\nsearch");
            searchValue = 30;
            Console.WriteLine($"  入力値: {searchValue}");
            output = arrayData.Search(searchValue);
            Console.WriteLine($"  出力値: {output}");

            Console.WriteLine("\nBinarySearch TEST <----- end");
        }
    }
}

File: array_search/search_binary/src/BinarySearchDemo.py
# Python
# 配列の検索: 二分探索 (Binary Search)

class ArrayData:
    def __init__(self):
        self._data = []

    def get(self):
        return self._data

    def set(self, data):
        self._data = data
        return True

    def search(self, target):
        left = 0
        right = len(self._data) - 1
        
        while left <= right:
            mid = (left + right) // 2
            
            # 中央の要素が目標値と一致
            if self._data[mid] == target:
                return mid
            
            # 中央の要素が目標値より小さい場合、右半分を探索
            elif self._data[mid] < target:
                left = mid + 1
            
            # 中央の要素が目標値より大きい場合、左半分を探索
            else:
                right = mid - 1
        
        # 目標値が見つからない場合
        return -1

def main():
    print("BinarySearch TEST -----> start")

    print("\nnew")
    array_data = ArrayData()
    input = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
    array_data.set(input)
    print(f"  現在のデータ: {array_data.get()}")
    
    print("\nsearch")
    input = 7
    print(f"  入力値: {input}")
    output = array_data.search(input)
    print(f"  出力値: {output}")

    print("\nsearch")
    input = 30
    print(f"  入力値: {input}")
    output = array_data.search(input)
    print(f"  出力値: {output}")

    print("\nBinarySearch TEST <----- end")

if __name__ == "__main__":
    main()